packages/slate-editor/src/extensions/placeholders/elements/AttachmentPlaceholderElement.tsx

```diff
@@ -41,7 +41,15 @@ export function AttachmentPlaceholderElement({ children, element, ...props }: Pr
                 const caption = fileInfo[UPLOADCARE_FILE_DATA_KEY]?.caption || '';
                 return { file: file.toPrezlyStoragePayload(), caption };
             });
-            PlaceholdersManager.register(element.type, placeholders[i].uuid, uploading);
+
+            // We need to delay registering additional placeholders to make sure they are mounted
+            // and the follower is created in `PlaceholdersManager` to respond to the `onResolve` callback.
+            setTimeout(
+                () => {
+                    PlaceholdersManager.register(element.type, placeholders[i].uuid, uploading);
+                },
+                i > 0 ? 100 : 0,
```

Comment:
100ms should be plenty enough for the components to mount. I tested with various amounts of images selected (from 2 to 10), and the mounting took at most 60ms to complete. The time was pretty consistent around 50ms, since all placeholders are inserted at once, so there's really only one re-render happening.

Reply:
What if we always delay notifications in case a new follower will come late? Would this be *slightly* more reliable than a 100ms timeout?



```diff
--- a/packages/slate-editor/src/extensions/placeholders/PlaceholdersManager.ts	(revision 6aefc3d3ad0f2ef92c2c01640d5764a6ae7c9fa5)
+++ b/packages/slate-editor/src/extensions/placeholders/PlaceholdersManager.ts	(date 1685529226441)
@@ -12,6 +12,8 @@
 
 import { PlaceholderNode } from './PlaceholderNode';
 
+const NOTIFICATION_TTL = 10000;
+
 const Type = PlaceholderNode.Type;
 type Type = PlaceholderNode.Type;
 type Uuid = PlaceholderNode['uuid'];
@@ -66,16 +68,24 @@
 type Follower<T extends Type> = Identifier<T> & Callbacks<T>;
 type Unfollow = () => void;
 
+type Notification<T extends Type> = {
+    type: T;
+    uuid: Uuid;
+    callback: (follower: Follower<T>) => void;
+};
+
 interface State {
     active: Identifier<Type> | undefined;
     followers: Array<Follower<Type>>;
     triggers: Identifier<Type>[];
+    notifications: Notification<Type>[];
 }
 
 const state: State = {
     active: undefined,
     followers: [],
     triggers: [],
+    notifications: [],
 };
 
 export const PlaceholdersManager = {
@@ -144,6 +154,7 @@
         state.followers = [...state.followers, follower];
 
         trigger(follower);
+        consumeNotifications(follower);
 
         return () => {
             state.followers = state.followers.filter((f) => f !== follower);
@@ -181,6 +192,13 @@
 }
 
 function notify<T extends Type>(type: T, uuid: Uuid, callback: (follower: Follower<T>) => void) {
+    const notification = { type, uuid, callback } as any as Notification<Type>;
+    state.notifications = [...state.notifications, notification];
+
+    setTimeout(() => {
+        state.notifications = state.notifications.filter((n) => n !== notification);
+    }, NOTIFICATION_TTL);
+
     state.followers.forEach((follower) => {
         if (is(follower, type, uuid)) {
             callback(follower as any as Follower<T>);
@@ -188,6 +206,14 @@
     });
 }
 
+function consumeNotifications<T extends Type>(follower: Follower<T>) {
+    state.notifications.forEach(({ type, uuid, callback }) => {
+        if (is(follower, type, uuid)) {
+            (callback as any as Notification<T>['callback'])(follower);
+        }
+    });
+}
+
 function trigger<T extends Type>({ type, uuid, onTrigger }: Follower<T>) {
     if (!onTrigger) return;
```

```diff
--- a/packages/slate-editor/src/extensions/placeholders/elements/ImagePlaceholderElement.tsx	(revision 6aefc3d3ad0f2ef92c2c01640d5764a6ae7c9fa5)
+++ b/packages/slate-editor/src/extensions/placeholders/elements/ImagePlaceholderElement.tsx	(date 1685529114877)
@@ -54,14 +54,7 @@
                 return { file: image.toPrezlyStoragePayload(), caption };
             });
 
-            // We need to delay registering additional placeholders to make sure they are mounted
-            // and the follower is created in `PlaceholdersManager` to respond to the `onResolve` callback.
-            setTimeout(
-                () => {
-                    PlaceholdersManager.register(element.type, placeholders[i].uuid, uploading);
-                },
-                i > 0 ? 100 : 0,
-            );
+            PlaceholdersManager.register(element.type, placeholders[i].uuid, uploading);
         });
     }
```


```diff
--- a/packages/slate-editor/src/extensions/placeholders/elements/AttachmentPlaceholderElement.tsx	(revision 6aefc3d3ad0f2ef92c2c01640d5764a6ae7c9fa5)
+++ b/packages/slate-editor/src/extensions/placeholders/elements/AttachmentPlaceholderElement.tsx	(date 1685529123749)
@@ -42,14 +42,7 @@
                 return { file: file.toPrezlyStoragePayload(), caption };
             });
 
-            // We need to delay registering additional placeholders to make sure they are mounted
-            // and the follower is created in `PlaceholdersManager` to respond to the `onResolve` callback.
-            setTimeout(
-                () => {
-                    PlaceholdersManager.register(element.type, placeholders[i].uuid, uploading);
-                },
-                i > 0 ? 100 : 0,
-            );
+            PlaceholdersManager.register(element.type, placeholders[i].uuid, uploading);
         });
     }
```

```diff
--- a/packages/slate-editor/src/extensions/placeholders/components/PlaceholderElement.tsx	(revision 6aefc3d3ad0f2ef92c2c01640d5764a6ae7c9fa5)
+++ b/packages/slate-editor/src/extensions/placeholders/components/PlaceholderElement.tsx	(date 1685528865763)
@@ -51,7 +51,7 @@
         element.type as PlaceholderNode.Type,
         element.uuid,
         {
-            onProgress: (p) => setProgress(p),
+            onProgress: setProgress,
         },
     );
```

Reply:
Wow, that's actually a really nice solution! 😮 
I've been focusing on the component mounting, and didn't even think tha changing the notifications behavior would be a feasible option.
I'll try your changes and get back with a result

Reply:
Hey @e1himself, I applied changes you proposed, and it worked flawlessly! 👏🏻 

See 4ba7f983f7200f638aa790aa97332ee0e4b4f111